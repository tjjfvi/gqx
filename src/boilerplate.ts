// @ts-nocheck
throw new Error("boilerplate.ts should not be imported")
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-unused-vars */

/* --- */

type Int = number;
type Float = number;
type String = string;
type ID = string;
type Boolean = boolean;

interface $<F, L> { f: F; l: L }
const $$ = <F, L>(f: F, l: L) => ({ f, l });

interface $$OperationId {
  typeProp: string;
  type: string;
  prop: string;
  inputTypes: { [x: string]: string };
}

const __wrap__ = Symbol();

type $$Wrap<X, Y> = X extends $<infer F, infer L> ? $<F, $$Wrap<L, Y>> : never;
type $$MapWrap<O, F> = {
  [K in keyof O | typeof __wrap__ | "$"]:
    K extends "$" ?
      <T extends $_>(x: T[]) => $<F, O extends { [__wrap__]: infer X } ? $$Wrap<X, T> : T>[] :
      K extends keyof O ?
        O[K] extends $_ ?
          $<F, O[K]> :
          $$MapWrap<O[K], F> :
        $<F, null>
}

const $$mapWrap = <O, F>(o: () => O, f: F): $$MapWrap<O, F> =>
  // @ts-ignore
  new Proxy(Object.create(null), {
    get: (t, k: string) => t[k] || (t[k] =
      k !== "$" ?
        o()[k].$ ?
          $$mapWrap(() => o()[k], f) :
          $$(f, o()[k]) :
        // @ts-ignore
        (a: any) => ("$" in o() ? o().$(a) : a).map((a: any) => $$(f, a))
    ),
  })

const $$reconstruct = <I extends $$OperationId>(id: I, input: $$Input.$<I>, props: $$Frag.$<I>[]) => {
  interface Subs { [k: string]: true | Subs }
  const subs: Subs = {};
  props.map(prop => populateSubs(prop, subs));
  const frag = genFrag(subs);
  const inputKeys = Object.keys(input);
  const inputDef = inputKeys.length ? `(${inputKeys.map(k =>
    `$${k}: ${id.inputTypes[k]}`
  ).join(", ")})` : "";
  const inputPass = inputKeys.length ? `(${inputKeys.map(k =>
    `${k}: $${k}`
  )})` : "";
  const request = `${id.typeProp}${inputDef} { ${id.prop}${inputPass}${frag} }`;
  return request;

  function genFrag(subs: Subs){
    return (
      Object.keys(subs).length ?
        `{ ${Object.entries(subs).map(([k, v]) => v === true ? k : k + " " + genFrag(v)).join(" ")} }` :
        ""
    );
  }

  function populateSubs(prop: $_, subs: Subs | true){
    if(subs === true)
      return;
    if("prop" in prop)
      subs[prop.prop] = true;
    else
      populateSubs(prop.l, subs[prop.f.prop] = (subs[prop.f.prop] || {}));
  }
}
