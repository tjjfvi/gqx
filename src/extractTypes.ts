import { DocumentNode, TypeNode } from "graphql"
import { Context } from "vm"
import { InputField, NamedType, Type, UnresolvedType } from "./type"

const baseScalars = ["ID", "String", "Int", "Float", "Boolean"]

export function extractTypes(documents: DocumentNode[]) {
  const types = new Map<string, NamedType>()

  for (const name of baseScalars) {
    types.set(name, {
      kind: "scalar",
      name,
      directives: [],
    })
  }

  const definitions = documents.flatMap(x => x.definitions)

  for (const typeDef of definitions) {
    if (typeDef.kind === "SchemaDefinition" || typeDef.kind === "SchemaExtension") {
      for (const operation of typeDef.operationTypes ?? []) {
        const type = getNamedType(operation.type.name.value)
        setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
        type.operationKey = operation.operation
      }
    }

    if (!("name" in typeDef) || !typeDef.name) continue

    const name = typeDef.name.value
    const type = getNamedType(name)

    if ("directives" in typeDef && typeDef.directives) {
      type.directives.push(...typeDef.directives)
    }

    if (typeDef.kind === "ScalarTypeDefinition" || typeDef.kind === "ScalarTypeExtension") {
      setKind(type, "scalar", {})
    } else if (
      typeDef.kind === "InputObjectTypeDefinition"
      || typeDef.kind === "InputObjectTypeExtension"
    ) {
      setKind(type, "input", { fields: [] })
      for (const fieldDef of typeDef.fields ?? []) {
        const key = fieldDef.name.value
        const fieldType = getType(fieldDef.type)
        type.fields.push({
          key,
          type: fieldType,
          directives: fieldDef.directives?.slice() ?? [],
        })
      }
    } else if (
      typeDef.kind === "ObjectTypeDefinition"
      || typeDef.kind === "ObjectTypeExtension"
      || typeDef.kind === "InterfaceTypeDefinition"
      || typeDef.kind === "InterfaceTypeExtension"
    ) {
      setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
      for (const fieldDef of typeDef.fields ?? []) {
        const key = fieldDef.name.value
        const fullType = getType(fieldDef.type)
        const namedType = getNamedTypeOfType(fullType)
        const name = `${type.name}$${key}`
        type.fields.push({
          str: JSON.stringify(name),
          name,
          base: type,
          key,
          namedType,
          fullType,
          directives: fieldDef.directives?.slice() ?? [],
          input: fieldDef.arguments?.map((arg): InputField => ({
            key: arg.name.value,
            type: getType(arg.type),
            directives: arg.directives?.slice() ?? [],
          })) ?? [],
        })
      }
      if ("interfaces" in typeDef) {
        for (const supertypeDef of typeDef.interfaces ?? []) {
          const supertype = getNamedType(supertypeDef.name.value)
          setKind(supertype, "object", { fields: [], subtypes: [], supertypes: [] })
          type.supertypes.push(supertype)
          supertype.subtypes.push(type)
        }
      }
    } else if (typeDef.kind === "UnionTypeDefinition" || typeDef.kind === "UnionTypeExtension") {
      setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
      for (const subtypeDef of typeDef.types ?? []) {
        const subtype = getNamedType(subtypeDef.name.value)
        setKind(subtype, "object", { fields: [], subtypes: [], supertypes: [] })
        type.subtypes.push(subtype)
        subtype.supertypes.push(type)
      }
    } else if (typeDef.kind === "EnumTypeDefinition" || typeDef.kind === "EnumTypeExtension") {
      setKind(type, "enum", { values: [] })
      for (const value of typeDef.values ?? []) {
        type.values.push({
          name: value.name.value,
          directives: value.directives?.slice() ?? [],
        })
      }
    } else {
      types.delete(name)
    }
  }

  return [...types.values()]

  function getNamedType(name: string): NamedType {
    const existing = types.get(name)
    if (existing !== undefined) {
      return existing
    }
    const value = { kind: "unresolved", name, directives: [] } as any as NamedType
    types.set(name, value)
    return value
  }

  function getType(node: TypeNode): Type {
    let nullable = true
    if (node.kind === "NonNullType") {
      node = node.type
      nullable = false
    }
    let type: Type
    if (node.kind === "NamedType") {
      type = getNamedType(node.name.value)
    } else {
      type = { kind: "array", inner: getType(node.type) }
    }
    if (nullable) {
      type = { kind: "nullable", inner: type }
    }
    return type
  }

  function setKind<K extends NamedType["kind"]>(
    type: NamedType | UnresolvedType,
    kind: K,
    init: Omit<Extract<NamedType, { kind: K }>, keyof UnresolvedType>,
  ): asserts type is Extract<NamedType, { kind: K }> {
    if ((type.kind === "unresolved") as boolean) {
      type.kind = kind
      Object.assign(type, init)
    }
    if (type.kind !== kind) throw new Error("Conflicting definitions for type " + type.name)
  }

  function getNamedTypeOfType(type: Type): NamedType {
    while (type.kind === "nullable" || type.kind === "array") {
      type = type.inner
    }
    return type
  }
}

export { Context }
