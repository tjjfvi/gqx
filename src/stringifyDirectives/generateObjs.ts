
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
