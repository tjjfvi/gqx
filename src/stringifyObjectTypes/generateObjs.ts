
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
