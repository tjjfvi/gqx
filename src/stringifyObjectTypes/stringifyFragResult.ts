
import { Context, Obj, Prop } from ".";
import stringifyId from "./stringifyId";

type P = [Prop, string, string];

export default (obj: Obj, ctx: Context) => {
  ctx.exports.push("$" + obj.type);
  const shallows = obj.shallowProps.map((prop): P =>
    [prop, "typeof " + stringifyId(prop.id), prop.type]
  );
  const deeps = obj.deepProps.map((prop): [Prop, (x: string) => string] =>
    [prop, x => `$<typeof ${stringifyId(prop.id)}, ${x}>`]
  ).map(([prop, sub]): P =>
    [prop, sub("any"), `$$${prop.type}<Extract<F, ${sub(prop.type + "$")}>["l"]>`]
  );
  const core = `type $$${obj.type}<F extends ${obj.type}$> =\n${[...shallows, ...deeps].map(x =>
    `  & (${x[1]} extends F ? { ${x[0].id.prop}: ${x[0].wrap(x[2])} } : {})\n`
  ).join("")}`;
  const alias = `type $${obj.type}<F extends ${obj.type}$[]> = $$${obj.type}<F[number]>;`;
  return core + alias;
}
