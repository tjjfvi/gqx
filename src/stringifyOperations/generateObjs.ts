
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
