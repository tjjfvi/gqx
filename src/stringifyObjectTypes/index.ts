
import { Context } from "..";
import stringifyFragEnums from "./stringifyFragEnums";
import stringifyFragResult from "./stringifyFragResult";
import { stringifyObjectTypeInfo } from "./stringifyObjectTypeInfo";
import stringifyFragObject from "./stringifyFragObject";
import generateObjs from "./generateObjs";

interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  shallow: boolean;
  wrap: (x: string) => string;
  wrapHKT: string;
}

interface Obj {
  type: string;
  shallowProps: Prop[];
  deepProps: Prop[];
}

export { Id, Prop, Obj, Context };

export default (ctx: Context) => {
  const [objs] = generateObjs(ctx);

  return [
    stringifyObjectTypeInfo(objs),
    stringifyFragEnums(objs),
    ...objs.map(o => stringifyFragObject(o)),
    ...objs.map(o => stringifyFragResult(o))
  ].join("\n\n");
}

