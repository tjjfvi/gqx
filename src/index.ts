
import {
  parse,
  InputValueDefinitionNode,
  FieldDefinitionNode,
  EnumValueDefinitionNode
} from "graphql";
import stringifyEnumTypes from "./stringifyEnumTypes";
import stringifyInputTypes from "./stringifyInputTypes";
import stringifyObjectTypes from "./stringifyObjectTypes";
import stringifyOperations from "./stringifyOperations";
import stringifyBoilerplate from "./stringifyBoilerplate";
import groupDefinitions from "./groupDefinitions";
import loadConfig, { Config } from "./loadConfig";
import stringifyDirectives from "./stringifyDirectives";

interface Context {
  inputTypes: { [k: string]: InputValueDefinitionNode[] };
  enumTypes: { [k: string]: EnumValueDefinitionNode[] };
  objectTypes: { [k: string]: FieldDefinitionNode[] };
  operations: [string, string][];
  baseTypes: string[];
  config: Config;
}

export default ({ schema, template }: { schema: string; template: string }) => {

  const ast = parse(schema);
  const [config, wrapper] = loadConfig(template);

  const ctx: Context = {
    operations: [],
    baseTypes: [],
    enumTypes: {},
    inputTypes: {},
    objectTypes: {},
    config,
  };

  groupDefinitions(ctx, ast.definitions);

  const code = [
    stringifyBoilerplate(),
    stringifyEnumTypes(ctx),
    stringifyInputTypes(ctx),
    stringifyObjectTypes(ctx),
    stringifyOperations(ctx),
    stringifyDirectives(ctx),
  ].join("\n\n").replace(/\n\n\n+/g, "\n\n");

  return wrapper(code);

}

export { Context };
