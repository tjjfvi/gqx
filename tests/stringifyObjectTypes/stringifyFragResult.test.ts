
import generateObjs from "../../src/stringifyObjectTypes/generateObjs"
import stringifyFragResult from "../../src/stringifyObjectTypes/stringifyFragResult";
import { Obj } from "../../src/stringifyObjectTypes";

describe("extensive.gql", () => {
  const [objs,, ctx] = generateObjs(utils.getOperations(["extensive"]));
  test.each(objs.map((o): [string, Obj] => [o.type, o]))("%s", (_, o) => {
    const code = stringifyFragResult(o, ctx);
    expect(code).toParse().toMatchSnapshot();
  })
})

