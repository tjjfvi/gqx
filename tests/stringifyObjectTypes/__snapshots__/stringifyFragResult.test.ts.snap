// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyObjectTypes.gql BinaryTree 1`] = `
"type _$BinaryTree<F extends BinaryTree$> =
  & (typeof BinaryTree$node extends F ? { node: String } : {})
  & ($<typeof BinaryTree$left, any> extends F ? { left: (_$BinaryTree<Extract<F, $<typeof BinaryTree$left, BinaryTree$>>[\\"l\\"]> | null) } : {})
  & ($<typeof BinaryTree$right, any> extends F ? { right: (_$BinaryTree<Extract<F, $<typeof BinaryTree$right, BinaryTree$>>[\\"l\\"]> | null) } : {})
export type $BinaryTree<F extends BinaryTree$[]> = _$BinaryTree<F[number]>;"
`;

exports[`stringifyObjectTypes.gql DeepOnly 1`] = `
"type _$DeepOnly<F extends DeepOnly$> =
  & ($<typeof DeepOnly$a, any> extends F ? { a: _$ShallowOnly<Extract<F, $<typeof DeepOnly$a, ShallowOnly$>>[\\"l\\"]> } : {})
  & ($<typeof DeepOnly$b, any> extends F ? { b: (_$ShallowOnly<Extract<F, $<typeof DeepOnly$b, ShallowOnly$>>[\\"l\\"]> | null)[] } : {})
export type $DeepOnly<F extends DeepOnly$[]> = _$DeepOnly<F[number]>;"
`;

exports[`stringifyObjectTypes.gql Recursive 1`] = `
"type _$Recursive<F extends Recursive$> =
  & (typeof Recursive$id extends F ? { id: ID } : {})
  & ($<typeof Recursive$next, any> extends F ? { next: _$Recursive<Extract<F, $<typeof Recursive$next, Recursive$>>[\\"l\\"]> } : {})
export type $Recursive<F extends Recursive$[]> = _$Recursive<F[number]>;"
`;

exports[`stringifyObjectTypes.gql ShallowOnly 1`] = `
"type _$ShallowOnly<F extends ShallowOnly$> =
  & (typeof ShallowOnly$a extends F ? { a: String } : {})
  & (typeof ShallowOnly$b extends F ? { b: Number } : {})
export type $ShallowOnly<F extends ShallowOnly$[]> = _$ShallowOnly<F[number]>;"
`;
